### ISSUE TITLE
[MENFORCER-426] DependencyConvergence in 3.1.0 fails when using version ranges

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#541]
**[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett)** opened **[MENFORCER-426](https://issues.apache.org/jira/browse/MENFORCER-426?redirect=false)** and commented

In our project, using version 3.0.0-M3 of the maven-enforcer-plugin's DependencyConvergence rule passes.  Using version 3.1.0 starts to show convergence errors that I believe may be related to using version ranges to declare dependencies:

```java
[WARNING] 
Dependency convergence error for com.trib3:graphql:jar:1.32.3279:compile paths to dependency are:
+-com.joe.quizzy:server:jar:1.0-SNAPSHOT
  +-com.joe.quizzy:graphql:jar:1.0-SNAPSHOT:compile
    +-com.trib3:graphql:jar:1.32.3279:compile
and
+-com.joe.quizzy:server:jar:1.0-SNAPSHOT
  +-com.trib3:graphql:jar:1.32.3324:compile
```

both `com.joe.quizzy:server` and `com.joe.quizzy:graphql` declare a dependency on `com.trib3:graphql`, with a version of `[1.32.1,1.33-SNAPSHOT)`.  that version range should (currently) resolve to 1.32.3324 in both usages, but in enforcer 3.1.0's DependencyConvergence check does not.

---

**Affects:** 3.0.0, 3.1.0

**Issue Links:**
- [MRESOLVER-345](https://issues.apache.org/jira/browse/MRESOLVER-345) Conflict resolution in verbose mode is sensitive to version ordering
 (_**"is caused by"**_)

**Remote Links:**
- [GitHub Pull Request #259
](https://github.com/apache/maven-enforcer/pull/259)

0 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett) - Can you provide a reproduce simple project?

**Comment by jira-importer** [comment]
**[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett)** commented

more minimal reproducer here: https://github.com/josephlbarnett/menforcer426-reproducer – commenting out the `version.maven.enforcer` property will use enforcer 3.0.0-M3 and `maven verify` will pass.  `maven verify` fails as is

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I hope also reproduce issue: https://github.com/apache/maven-enforcer/pull/259

**Comment by jira-importer** [comment]
**[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett)** commented

I still see a convergence error when running `mvn verify -Daether.dependencyCollector.impl=bf` in the linked reproducer projects

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

You need Maven 3.9.x BF is new implementation not present in older version.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Slawek created reproducer for this use case and had following results (running same IT w/ different enforcer version):
* 3.2.2-SNAPSHOT (current master) – FAILS
* 3.2.1 (uses resolver) -- FAILS
* 3.1.0 (uses old shared-dependency-tree) – FAILS
* 3.0.0 (uses old shared-dependency-tree) – FAILS
* 3.0.0-M3 (uses older shared-dependency-tree) – PASSES OK

Hence, it seems that it is this commit (that is present in 3.0.0 and upwards, but not in 3.0.0-M3) https://github.com/apache/maven-enforcer/commit/ca40308fd58c45e638a35768b3966b5680d4c60e that causes this issue to become visible.

To make things more interesting, we do have a bug in "original" DF resolver itself, recorded as MRESOLVER-345 that just complicates issue investigation more (as this bug is present in all resolver versions, moreover in pre 3.9.0 Maven versions only DF was present).

For this we need some information from reporter [Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett) :
* what Maven version you use?
* for statement "using version 3.0.0-M3 of the maven-enforcer-plugin's DependencyConvergence rule passes.  Using version 3.1.0 starts to show convergence errors" you hopefully used same Maven version (that in bullet above)?

**Comment by jira-importer** [comment]
**[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) have been using maven versions 3.6.3 and 3.8.5 for this, and seen the same behavior in both versions.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

As [sjaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sjaranowski) mentioned, one workaround (that works ONLY with Maven 3.9.1 would be to invoke your build using Maven 3.9.1 and with `-Daether.dependencyCollector.impl=bf`

**Comment by jira-importer** [comment]
**[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett)** commented

can confirm that workaround does work on 3.9.1 (but we're more likely to stay on enforcer 3.0.0-M3 than add that workaround at this point in time)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Joe Barnett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbarnett) Thanks you a LOT for confirmation.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

Another workaround is to exclude problematic artifacts:
In your case configuration can look like:

```
<dependencyConvergence>
  <excludes>
    <exclude>com.trib3:graphql</exclude>
  </excludes>
</dependencyConvergence>
```

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I fixed my reproduce test to execute only on Maven 3.9.2 and we see that is ok.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I close as we can not do more in plugin, by the way issue was fixed in Maven 3.9.2
